File: averis/lib/care/escalation.ts
/**
 * Escalation — when a finding becomes somebody's problem.
 *
 * An alert says a measurement crossed a threshold. An emergency event says a
 * human needs to respond, and it appears in a clinician's queue until one
 * does. The distance between those two claims is this file.
 *
 * ── Why this is rules and not a model ──────────────────────────────────────
 *
 * The same reasoning as the threshold alerts and the Phase 3 insight engine:
 * an escalation wakes a person up, and the person woken deserves to know
 * exactly what tripped it. "SpO2 was 86% and the escalation threshold is 90%"
 * is checkable at 3am; "the model was confident" is not.
 *
 * The AI engine can raise one — `RAPID_DETERIORATION` exists precisely so a
 * trend nobody's threshold catches can still reach a clinician — but even then
 * the *decision* is a rule applied to the model's output, and the evidence
 * travels with the event.
 *
 * ── The two suppressions, and why they are not the same ────────────────────
 *
 * `shouldEscalate` suppresses a repeat of something already open. This is not
 * the alert-level suppression repeated at a second layer: alerts dedupe so the
 * patient's alert list stays readable, and emergencies dedupe so a clinician's
 * response queue does not fill with the same unanswered event 300 times while
 * they are on their way to the patient. Both matter; neither substitutes for
 * the other, because an acknowledged emergency stops suppressing new *alerts*
 * and an open emergency must keep suppressing new *emergencies*.
 *
 * Pure — no database, no clock beyond what callers pass in. The runtime that
 * actually raises these is the Python ingest service, and `escalation.py` is a
 * port of this file covered by the same expectations.
 */

export type EmergencyType =
  | "FALL_DETECTED"
  | "SEVERE_HYPOXIA"
  | "EXTREME_HEART_RATE"
  | "RAPID_DETERIORATION"
  | "DEVICE_LOST"
  | "MANUAL_ESCALATION";

export type EmergencySeverity = "INFO" | "WARNING" | "CRITICAL";

export type EmergencyStatus =
  | "NEW"
  | "ACKNOWLEDGED"
  | "IN_PROGRESS"
  | "RESOLVED"
  | "DISMISSED";

export type DetectedBy = "RULE_ENGINE" | "AI_ENGINE" | "PATIENT" | "CLINICIAN";

/** A candidate escalation, before anything has been written down. */
export type EmergencyCandidate = {
  eventType: EmergencyType;
  severity: EmergencySeverity;
  detectedBy: DetectedBy;
  /** One sentence a clinician can act on without opening the chart. */
  summary: string;
  /**
   * What the engine saw. An emergency a clinician cannot trace back to numbers
   * is one they have to take on trust at the worst possible moment.
   */
  evidence: Record<string, unknown>;
};

/** The subset of an alert this module needs. */
export type AlertLike = {
  alertType: string;
  severity: EmergencySeverity;
  message: string;
  observedValue: number | null;
  thresholdValue: number | null;
};

/** The subset of an AI assessment this module needs. */
export type AssessmentLike = {
  riskLevel: "LOW" | "MODERATE" | "HIGH" | "CRITICAL" | null;
  riskScore: number | null;
  confidence: number | null;
  /** Human-readable reasons, already produced by the engine. */
  reasons: string[];
  /** Whether the fall model fired on the most recent IMU window. */
  fallDetected: boolean;
  fallConfidence: number | null;
  /** Direction of travel over the assessment window, if the engine reported one. */
  deteriorating: boolean;
};

export type OpenEmergency = {
  eventType: EmergencyType;
  severity: EmergencySeverity;
  status: EmergencyStatus;
};

/**
 * Risk score at which the AI engine may escalate on its own.
 *
 * 0.85 rather than 0.7, and CRITICAL rather than HIGH, because this threshold
 * buys a clinician's attention with a model's opinion. Set it low and the
 * queue fills with assessments nobody can check; set it too high and the
 * trend-only deterioration the thresholds cannot see never reaches anyone.
 */
export const AI_ESCALATION_SCORE = 0.85;

/**
 * How long a device may stay silent before that is itself the emergency.
 *
 * Fifteen minutes, not one: bands drop off Wi-Fi, phones move out of
 * Bluetooth range, and a system that escalates on every gap teaches its
 * clinicians to dismiss escalations. Fifteen minutes of nothing from a device
 * that was reporting is no longer a network blip.
 */
export const DEVICE_SILENCE_MS = 15 * 60 * 1000;

const SEVERITY_RANK: Record<EmergencySeverity, number> = {
  INFO: 1,
  WARNING: 2,
  CRITICAL: 3,
};

/** Statuses that mean nobody has finished dealing with it yet. */
export const OPEN_STATUSES: EmergencyStatus[] = ["NEW", "ACKNOWLEDGED", "IN_PROGRESS"];

/**
 * Which threshold alerts are emergencies in their own right.
 *
 * A WARNING never appears here. The whole point of two levels is that one of
 * them can be looked at later, and an escalation path that treats them alike
 * produces a queue with no priority in it.
 */
export function fromAlerts(alerts: AlertLike[]): EmergencyCandidate[] {
  const candidates: EmergencyCandidate[] = [];

  for (const alert of alerts) {
    if (alert.severity !== "CRITICAL") continue;

    if (alert.alertType === "FALL_SUSPECTED") {
      candidates.push({
        eventType: "FALL_DETECTED",
        severity: "CRITICAL",
        detectedBy: "RULE_ENGINE",
        summary: "The device reported a movement pattern consistent with a fall.",
        evidence: { alertType: alert.alertType, source: "movement_status" },
      });
      continue;
    }

    if (alert.alertType === "SPO2_LOW") {
      candidates.push({
        eventType: "SEVERE_HYPOXIA",
        severity: "CRITICAL",
        detectedBy: "RULE_ENGINE",
        summary: alert.message,
        evidence: {
          alertType: alert.alertType,
          observed: alert.observedValue,
          threshold: alert.thresholdValue,
        },
      });
      continue;
    }

    if (alert.alertType === "HEART_RATE_HIGH" || alert.alertType === "HEART_RATE_LOW") {
      candidates.push({
        eventType: "EXTREME_HEART_RATE",
        severity: "CRITICAL",
        detectedBy: "RULE_ENGINE",
        summary: alert.message,
        evidence: {
          alertType: alert.alertType,
          observed: alert.observedValue,
          threshold: alert.thresholdValue,
        },
      });
    }

    // TEMPERATURE_* is deliberately absent. A critical temperature is a real
    // finding and raises an alert, but it is not a minutes-matter event in the
    // way hypoxia, a collapse or an arrhythmia are, and putting it in the same
    // queue would dilute the queue rather than improve the temperature.
  }

  return dedupeByType(candidates);
}

/**
 * Whether the AI assessment is itself an escalation.
 *
 * Two ways in, and they are different claims:
 *
 *   · the fall model fired — a discrete event, like a threshold crossing
 *   · risk is critical *and* rising — a trend, which is the one thing the
 *     thresholds structurally cannot see, since every individual reading in a
 *     slow decline can sit inside the normal band
 *
 * A high score that is flat does not escalate. A patient who has been at 0.86
 * for a week is a patient whose clinician already knows.
 */
export function fromAssessment(assessment: AssessmentLike): EmergencyCandidate[] {
  const candidates: EmergencyCandidate[] = [];

  if (assessment.fallDetected) {
    candidates.push({
      eventType: "FALL_DETECTED",
      severity: "CRITICAL",
      detectedBy: "AI_ENGINE",
      summary: "The fall model detected a movement pattern consistent with a fall.",
      evidence: {
        model: "fall_detector",
        confidence: assessment.fallConfidence,
        // Named in the evidence because the model is trained on synthetic
        // motion, and a clinician reading this should know that without
        // having to find the model card.
        caveat: "trained on synthetic motion data",
      },
    });
  }

  const score = assessment.riskScore ?? 0;
  if (
    assessment.riskLevel === "CRITICAL" &&
    score >= AI_ESCALATION_SCORE &&
    assessment.deteriorating
  ) {
    candidates.push({
      eventType: "RAPID_DETERIORATION",
      severity: "CRITICAL",
      detectedBy: "AI_ENGINE",
      summary:
        assessment.reasons.length > 0
          ? `Risk assessment reached ${Math.round(score * 100)}% and is rising: ${assessment.reasons
              .slice(0, 3)
              .join("; ")}.`
          : `Risk assessment reached ${Math.round(score * 100)}% and is rising.`,
      evidence: {
        riskScore: score,
        riskLevel: assessment.riskLevel,
        confidence: assessment.confidence,
        reasons: assessment.reasons.slice(0, 5),
      },
    });
  }

  return dedupeByType(candidates);
}

/**
 * A device that has stopped reporting.
 *
 * WARNING, not CRITICAL. Nothing is known to be wrong with the patient — that
 * is precisely the problem, and it is a different problem from a measured
 * emergency. Ranking a lost signal alongside a fall would teach clinicians
 * that red does not mean red.
 *
 * A device that was never connected does not escalate: `lastReadingAt` of null
 * is a band still in its box, not a band that went quiet.
 */
export function fromSilence(
  lastReadingAt: string | null,
  now: number,
  silenceMs = DEVICE_SILENCE_MS,
): EmergencyCandidate[] {
  if (!lastReadingAt) return [];

  const last = Date.parse(lastReadingAt);
  if (Number.isNaN(last)) return [];

  const silent = now - last;
  if (silent < silenceMs) return [];

  const minutes = Math.floor(silent / 60000);
  return [
    {
      eventType: "DEVICE_LOST",
      severity: "WARNING",
      detectedBy: "RULE_ENGINE",
      summary: `No readings for ${minutes} minutes. This patient is not currently being monitored.`,
      evidence: { lastReadingAt, silentMinutes: minutes },
    },
  ];
}

/**
 * Whether a candidate should be written down given what is already open.
 *
 * Mirrors the partial unique index on `emergency_events` — one open event per
 * patient per type — so the application refuses before the database has to,
 * and a caller gets a decision rather than a constraint violation.
 *
 * An escalation still gets through: an open WARNING for a device that has now
 * gone CRITICAL is new information, and the queue must be able to say so.
 */
export function shouldEscalate(
  candidate: EmergencyCandidate,
  open: OpenEmergency[],
): boolean {
  const existing = open.find(
    (e) => e.eventType === candidate.eventType && OPEN_STATUSES.includes(e.status),
  );
  if (!existing) return true;

  return SEVERITY_RANK[candidate.severity] > SEVERITY_RANK[existing.severity];
}

/**
 * Everything a reading and its assessment justify raising, already deduped
 * against what is open.
 *
 * The order matters for the caller that writes them: most severe first, so a
 * partial failure loses the least important event rather than the one that
 * mattered.
 */
export function escalationsFor(input: {
  alerts?: AlertLike[];
  assessment?: AssessmentLike | null;
  lastReadingAt?: string | null;
  now?: number;
  open?: OpenEmergency[];
}): EmergencyCandidate[] {
  const candidates = dedupeByType([
    ...fromAlerts(input.alerts ?? []),
    ...(input.assessment ? fromAssessment(input.assessment) : []),
    ...(input.lastReadingAt !== undefined
      ? fromSilence(input.lastReadingAt, input.now ?? Date.now())
      : []),
  ]);

  return candidates
    .filter((candidate) => shouldEscalate(candidate, input.open ?? []))
    .sort((a, b) => SEVERITY_RANK[b.severity] - SEVERITY_RANK[a.severity]);
}

/**
 * One event per type, keeping the most severe.
 *
 * The rule engine and the AI engine can both report a fall from the same
 * instant — one from the movement flag, one from the model. That is one fall,
 * and a queue showing it twice makes a clinician check whether the patient
 * fell twice.
 */
function dedupeByType(candidates: EmergencyCandidate[]): EmergencyCandidate[] {
  const best = new Map<EmergencyType, EmergencyCandidate>();

  for (const candidate of candidates) {
    const existing = best.get(candidate.eventType);
    if (!existing || SEVERITY_RANK[candidate.severity] > SEVERITY_RANK[existing.severity]) {
      best.set(candidate.eventType, candidate);
    }
  }

  return [...best.values()];
}

/** Plain-language labels, shared by the queue, the inbox and the assistant. */
export const EMERGENCY_LABEL: Record<EmergencyType, string> = {
  FALL_DETECTED: "Fall detected",
  SEVERE_HYPOXIA: "Severe drop in blood oxygen",
  EXTREME_HEART_RATE: "Extreme heart rate",
  RAPID_DETERIORATION: "Rapidly rising risk",
  DEVICE_LOST: "Device stopped reporting",
  MANUAL_ESCALATION: "Escalated by hand",
};

export type CareNotice = {
  title: string;
  body: string;
  /** Relative, always. An absolute URL here is an open redirect people trust. */
  href: string;
  severity: EmergencySeverity;
};

/**
 * What a care team member is actually told.
 *
 * The patient's name goes in the title and the finding in the body, in that
 * order, because a phone notification truncates and the recipient's first
 * question is always *who*. Everything after the first line is detail they
 * will read on the chart.
 *
 * No measurements beyond the ones already in the summary, and never a
 * suggestion of what to do about it: the notice's job is to get a clinician to
 * the patient, and the chart's job is to inform what happens next.
 */
export function noticeFor(
  candidate: Pick<EmergencyCandidate, "eventType" | "severity" | "summary">,
  patient: { patientId: string; fullName: string },
): CareNotice {
  return {
    title: `${patient.fullName} — ${EMERGENCY_LABEL[candidate.eventType]}`,
    body: candidate.summary,
    href: `/clinical/${patient.patientId}`,
    severity: candidate.severity,
  };
}

File: averis/iot-service/app/escalation.py
"""Escalation — when a finding becomes somebody's problem.

An alert says a measurement crossed a threshold. An emergency event says a
human needs to respond, and it stays in a clinician's queue until one does.

Ported from `lib/care/escalation.ts`, which carries the full reasoning. The
short version:

  · rules, not a model, because an escalation wakes a person up and the person
    woken deserves to know exactly what tripped it
  · the AI engine may still raise one, since a slow decline in which every
    individual reading sits inside the normal band is the one thing thresholds
    structurally cannot see — but only when risk is critical *and* rising
  · a critical temperature does not escalate; diluting the response queue would
    not improve the fever

This is the copy that runs in production. Keeping both is the same arrangement
as `alerts.py` and `lib/iot/alert-rules.ts`: one language owns the ingest path,
the other owns what the dashboard reasons about, and they are tested against
the same expectations.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone

from .alerts import Alert

SEVERITY_RANK = {"INFO": 1, "WARNING": 2, "CRITICAL": 3}

OPEN_STATUSES = ("NEW", "ACKNOWLEDGED", "IN_PROGRESS")

# 0.85 rather than 0.7, and CRITICAL rather than HIGH: this threshold buys a
# clinician's attention with a model's opinion.
AI_ESCALATION_SCORE = 0.85

# Fifteen minutes. Bands drop off Wi-Fi and phones move out of range; a system
# that escalates on every gap teaches its clinicians to dismiss escalations.
DEVICE_SILENCE_SECONDS = 15 * 60

LABELS = {
    "FALL_DETECTED": "Fall detected",
    "SEVERE_HYPOXIA": "Severe drop in blood oxygen",
    "EXTREME_HEART_RATE": "Extreme heart rate",
    "RAPID_DETERIORATION": "Rapidly rising risk",
    "DEVICE_LOST": "Device stopped reporting",
    "MANUAL_ESCALATION": "Escalated by hand",
}

# Insight kinds that mean "moving in the wrong direction". A rising heart rate
# and a falling SpO2 are both deterioration; the engine has already decided the
# slope was steep enough to be worth reporting.
DETERIORATING_INSIGHTS = ("TREND_RISE", "TREND_DECLINE", "PATTERN_CORRELATION")


@dataclass(frozen=True)
class Emergency:
    event_type: str
    severity: str
    detected_by: str
    summary: str
    evidence: dict = field(default_factory=dict)

    @property
    def title_for(self) -> str:
        return LABELS.get(self.event_type, self.event_type)


@dataclass(frozen=True)
class Assessment:
    """The subset of a HealthAssessment escalation needs."""

    risk_level: str | None
    risk_score: float | None
    confidence: float | None
    reasons: list[str]
    fall_detected: bool
    fall_confidence: float | None
    deteriorating: bool


def assessment_from_dict(payload: dict) -> Assessment:
    """Adapts the engine's output without the engine knowing about escalation.

    `deteriorating` is derived from the insights rather than added to the
    engine, because "is this getting worse" is a question about the trend
    insights it already produces — and a new field on the assessment would be
    a second place for the same judgement to live.
    """
    fall = payload.get("fall") or {}
    insights = payload.get("insights") or []

    return Assessment(
        risk_level=payload.get("risk_level"),
        risk_score=payload.get("risk_score"),
        confidence=payload.get("confidence"),
        reasons=list(payload.get("explanation") or []),
        fall_detected=bool(fall.get("fall_detected")),
        fall_confidence=fall.get("confidence"),
        deteriorating=any(
            i.get("insight_type") in DETERIORATING_INSIGHTS
            and i.get("severity") in ("WARNING", "CRITICAL")
            for i in insights
        ),
    )


def from_alerts(alerts: list[Alert]) -> list[Emergency]:
    """Which threshold alerts are emergencies in their own right.

    A WARNING never appears here. The point of two levels is that one of them
    can be looked at later.
    """
    out: list[Emergency] = []

    for alert in alerts:
        if alert.severity != "CRITICAL":
            continue

        if alert.alert_type == "FALL_SUSPECTED":
            out.append(
                Emergency(
                    "FALL_DETECTED",
                    "CRITICAL",
                    "RULE_ENGINE",
                    "The device reported a movement pattern consistent with a fall.",
                    {"alert_type": alert.alert_type, "source": "movement_status"},
                )
            )
        elif alert.alert_type == "SPO2_LOW":
            out.append(
                Emergency(
                    "SEVERE_HYPOXIA",
                    "CRITICAL",
                    "RULE_ENGINE",
                    alert.message,
                    {
                        "alert_type": alert.alert_type,
                        "observed": alert.observed_value,
                        "threshold": alert.threshold_value,
                    },
                )
            )
        elif alert.alert_type in ("HEART_RATE_HIGH", "HEART_RATE_LOW"):
            out.append(
                Emergency(
                    "EXTREME_HEART_RATE",
                    "CRITICAL",
                    "RULE_ENGINE",
                    alert.message,
                    {
                        "alert_type": alert.alert_type,
                        "observed": alert.observed_value,
                        "threshold": alert.threshold_value,
                    },
                )
            )

        # TEMPERATURE_* is deliberately absent — see the TypeScript original.

    return _dedupe(out)


def from_assessment(assessment: Assessment) -> list[Emergency]:
    """Whether the AI assessment is itself an escalation.

    Two ways in: the fall model fired (a discrete event), or risk is critical
    and rising (a trend). A high score that is flat belongs to a clinician who
    already knows.
    """
    out: list[Emergency] = []

    if assessment.fall_detected:
        out.append(
            Emergency(
                "FALL_DETECTED",
                "CRITICAL",
                "AI_ENGINE",
                "The fall model detected a movement pattern consistent with a fall.",
                {
                    "model": "fall_detector",
                    "confidence": assessment.fall_confidence,
                    # In the evidence because the model is trained on synthetic
                    # motion, and a clinician reading this should not have to
                    # find the model card to learn that.
                    "caveat": "trained on synthetic motion data",
                },
            )
        )

    score = assessment.risk_score or 0.0
    if (
        assessment.risk_level == "CRITICAL"
        and score >= AI_ESCALATION_SCORE
        and assessment.deteriorating
    ):
        percent = round(score * 100)
        reasons = assessment.reasons[:3]
        summary = (
            f"Risk assessment reached {percent}% and is rising: {'; '.join(reasons)}."
            if reasons
            else f"Risk assessment reached {percent}% and is rising."
        )
        out.append(
            Emergency(
                "RAPID_DETERIORATION",
                "CRITICAL",
                "AI_ENGINE",
                summary,
                {
                    "risk_score": score,
                    "risk_level": assessment.risk_level,
                    "confidence": assessment.confidence,
                    "reasons": assessment.reasons[:5],
                },
            )
        )

    return _dedupe(out)


def from_silence(
    last_reading_at: str | datetime | None,
    now: datetime | None = None,
    silence_seconds: int = DEVICE_SILENCE_SECONDS,
) -> list[Emergency]:
    """A device that has stopped reporting.

    WARNING, not CRITICAL. Nothing is known to be wrong with the patient —
    which is exactly the problem, and a different problem from a measured
    emergency. A device that never reported does not escalate: that is a band
    still in its box, not one that went quiet.
    """
    if last_reading_at is None:
        return []

    now = now or datetime.now(timezone.utc)

    if isinstance(last_reading_at, str):
        try:
            last = datetime.fromisoformat(last_reading_at.replace("Z", "+00:00"))
        except ValueError:
            return []
    else:
        last = last_reading_at

    if last.tzinfo is None:
        last = last.replace(tzinfo=timezone.utc)

    silent = (now - last).total_seconds()
    if silent < silence_seconds:
        return []

    minutes = int(silent // 60)
    return [
        Emergency(
            "DEVICE_LOST",
            "WARNING",
            "RULE_ENGINE",
            f"No readings for {minutes} minutes. "
            "This patient is not currently being monitored.",
            {"last_reading_at": last.isoformat(), "silent_minutes": minutes},
        )
    ]


def should_escalate(candidate: Emergency, open_events: list[dict]) -> bool:
    """Whether to write it down given what is already open.

    Mirrors the partial unique index on `emergency_events`, so the caller gets
    a decision rather than a constraint violation. An escalating severity still
    gets through: an open WARNING that has become CRITICAL is new information.
    """
    for existing in open_events:
        if existing.get("event_type") != candidate.event_type:
            continue
        if existing.get("status") not in OPEN_STATUSES:
            continue
        return SEVERITY_RANK[candidate.severity] > SEVERITY_RANK.get(
            existing.get("severity", "INFO"), 0
        )
    return True


def escalations_for(
    alerts: list[Alert] | None = None,
    assessment: Assessment | None = None,
    last_reading_at: str | datetime | None = None,
    now: datetime | None = None,
    open_events: list[dict] | None = None,
) -> list[Emergency]:
    """Everything a reading and its assessment justify, already deduped.

    Ordered most severe first, so a partial failure loses the least important
    event rather than the one that mattered.
    """
    candidates = _dedupe(
        [
            *from_alerts(alerts or []),
            *(from_assessment(assessment) if assessment else []),
            *(from_silence(last_reading_at, now) if last_reading_at else []),
        ]
    )

    survivors = [c for c in candidates if should_escalate(c, open_events or [])]
    return sorted(survivors, key=lambda c: SEVERITY_RANK[c.severity], reverse=True)


def notice_title(candidate: Emergency, patient_name: str) -> str:
    """What the care team member's notification says first.

    Name then finding: a phone notification truncates, and the recipient's
    first question is always *who*.
    """
    return f"{patient_name} — {candidate.title_for}"


def _dedupe(candidates: list[Emergency]) -> list[Emergency]:
    """One event per type, keeping the most severe.

    The rule engine and the fall model can both report the same fall. That is
    one fall, and a queue showing it twice makes a clinician check whether the
    patient fell twice.
    """
    best: dict[str, Emergency] = {}

    for candidate in candidates:
        existing = best.get(candidate.event_type)
        if existing is None or SEVERITY_RANK[candidate.severity] > SEVERITY_RANK[existing.severity]:
            best[candidate.event_type] = candidate

    return list(best.values())
